sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileIO.java

```diff
@@ -738,36 +772,12 @@ private ToReadableFileFn(ReadMatches spec) {
 
       @ProcessElement
       public void process(ProcessContext c) {
-        MatchResult.Metadata metadata = c.element();
-        if (metadata.resourceId().isDirectory()) {
-          switch (spec.getDirectoryTreatment()) {
-            case SKIP:
-              return;
-
-            case PROHIBIT:
-              throw new IllegalArgumentException(
-                  "Trying to read " + metadata.resourceId() + " which is a directory");
-
-            default:
-              throw new UnsupportedOperationException(
-                  "Unknown DirectoryTreatment: " + spec.getDirectoryTreatment());
-          }
+        ReadableFile r =
+            matchToReadableFile(c.element(), spec.getDirectoryTreatment(), spec.getCompression());
+        if (r == null) {
```

Comment:
Without reading everything it's unclear why match can be a `null`, e.g. there could be an error of some sort (access permission), or no matches or something else. I suggest creating another method along the lines of:

```java
boolean shouldSkipDirectory(metadata) {
  if (... isDirectory()) {
     ..
     case SKIP:
       return true;
     case PROHIBIT:
       throw ...
  }
}
```

This way it's clear that the point is specifically to handle/skip directories:

```java

void process(c) {
  if (shouldSkipDirectory(c.element())) {
    return;
  }
  c.output(toReadableFile(c.element()));
}